### DISCUSSION TITLE
Request for Notifier

### DISCUSSION THREAD

**Discussion by daphilli224** [zencart/zencart#6872] [Ideas]
Please consider adding a notifier in functions_osh_update.php around line 150 to give an observer a chance to modify the email to the customer. Please include $orders_id, $email_subject, $email_text, $html_msg, and possibly $notify_customer to be passed to the observer. Thank you.

**Comment by lat9** [comment]
What values would you expect an observer to be able to modify?

**Comment by daphilli224** [comment]
Definitely $html_msg and $email_text. I’ve modified the core ‘functions_osh_update.php' to add a custom notifier I’ve named ZEN_UPDATE_ORDERS_HISTORY_BEFORE_SENDING_CUSTOMER_EMAIL. In the observer, I change the customer name to only the customer’s first name to be a bit more friendly. I reorder and modify some of the elements of the email to be more in line with what I believe is a better order. And I add a note element to the html and text email versions.

The other suggestions for data passed to the observer are guessing what other data might be modified by other developers.

Dave

**Comment by drbyte** [comment]
Could the existing support for replaceable placeholders within the html email template be leveraged to do this already?

ie:
- add a note-element placeholder
- rearrange other content to reflect the desired alternate order
- and then when the text-only portion is built, it inherits all those already

**Comment by daphilli224** [comment]
Chris, I’m not sure I know what the replaceable placeholders support is that you mentioned. Can you explain?

I’ve edited the email template to get the html elements in the order I want, and created new and changed existing html elements in the observer mentioned above. The only way I know to change the html elements before the email is sent is to use an observer triggered by the requested notifier. And I need to modify the text email too.

Dave

**Comment by drbyte** [comment]
After looking at code, yes, you'll need a notifier that allows `$html_msg` to be edited by the observer.

`$html_msg` is an array of placeholders that get substituted when `zen_mail()` is called, as long as they follow the pattern of `$` followed by the placeholder name. 

ie: setting `$html_msg['ADDED_NOTES']` would replace `$ADDED_NOTES` if found in the `email_whatever.html` template's HTML when assembling all the message pieces.

Thus, rearranging the email's html template for your needs, and adding additional sections, could be done with those 2 parts ... to potentially make your custom email content preparation simpler, instead of rebuilding the whole thing in your observer.

I'm not opposed to the other variables you proposed having access to.
